EIPS/eip-7002.md

```diff
@@ -0,0 +1,380 @@
+
+    new_excess_exits = 0
+    if previous_excess_exits + exit_count > TARGET_EXITS_PER_BLOCK:
+        new_excess_exits = previous_excess_exits + exit_count - TARGET_EXITS_PER_BLOCK
+    
+    sstore(EXIT_PRECOMPILE_ADDRESS, EXCESS_EXITS_STORAGE_SLOT, new_excess_exits)
+    
+def reset_exit_count():
+    sstore(EXIT_PRECOMPILE_ADDRESS, EXIT_COUNT_STORAGE_SLOT, 0)
+```
+
+
+### Consensus layer
+
+In progress.
+
+Sketch of spec:
+
+* New operation `ExecutionLayerExit`
+* Will show up in `ExecutionPayload` as an SSZ List bound by length `MAX_EXITS_PER_BLOCK`
+* New function in `process_execution_layer_exit` that has similar functionality to `process_voluntary_exit` but that can fail validations (e.g. validator is already exited) without the block failing (similar to deposit coming from EL)
+* `process_execution_layer_exit` called in `process_operations` for each `ExecutionLayerExit` found in the `ExecutionPayload`
+
+## Rationale
+
+### Stateful precompile
+
+This specification utilizes a *stateful* precompile for simplicity and future-proofness. While precompiles are a well-known quantity, none to date have associated EVM state at the address.
+
+The alternative designs are (1) to utilize a precompile or opcode for the functionality and write a separate specified space in the EVM -- e.g. `0xFF..FF` -- or (2) to place the required state into the block and require the previous block header as an input into the state transition function (e.g. like [EIP-1559](./eip-1559.md) `base_fee`).
+
+Alternative design (1) is essentially using a stateful precompile but dissociating the state into a separate address. At first glance, this split appears unnecessarily convoluted when we could store the location of the `CALL` and the associated state in the same address. That said, there might be unexpected engineering constraints around precompiles in existing clients that make this a preferable path.
+
+Alternative design (2) has two main drawbacks. The first is that with the message queue contains an unbounded amount of state (as opposed to simple the `base_fee` in the similar EIP-1559 design). Additionally, even if the state was constrained to a single variable or two, this design pattern reinforces that the Ethereum state transition function signature be more than `f(pre_state, block) -> post_state` by putting another dependency on the `pre_block_header`. These additional dependencies hinder the elegance of future stateless designs. Providing these dependencies within the EVM state as specified, allows for them to show up naturally in block witnesses.
+
+### `validator_pubkey` field
+
+Multiple validators can utilize the same execution layer withdrawal credential, thus the `validator_pubkey` field is utilized to disambiguate which validator is being exited.
+
+Note, `validator_index` also disambiguates validators but is not used because the execution-layer cannot currently trustlessly ascertain this value.
```

Comment:
would you re-work this now that 4788 is shipping in Cancun?

I've always been a pubkey person instead of an index person but any reuse scenarios should effect this if we demand a proof during the flow of EL-exit creation